sdk/core/core-lro/src/lroEngine/azureAsyncPolling.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  failureStates,
+  LroResourceLocationConfig,
+  LongRunningOperation,
+  LroBody,
+  LroResponse,
+  LroStatus,
+  RawResponse,
+  successStates
+} from "./models";
+import { isUnexpectedPollingResponse } from "./requestUtils";
+
+function getResponseStatus(rawResponse: RawResponse): string {
+  const { status } = (rawResponse.body as LroBody) ?? {};
+  return status?.toLowerCase() ?? "succeeded";
+}
+
+function isAzureAsyncPollingDone(rawResponse: RawResponse): boolean {
+  const state = getResponseStatus(rawResponse);
+  if (isUnexpectedPollingResponse(rawResponse) || failureStates.includes(state)) {
+    throw new Error(`The long running operation has failed. The provisioning state: ${state}.`);
+  }
+  return successStates.includes(state);
```

Comment:
I'm a little curious about why failure is exceptional instead of being viewed as a final state, like success. Is failure never expected to happen normally, so this method is really just disambiguating between success and in progress?

Reply:
The code gen crew agreed on throwing if the operation reached a failure state and this function just implements this expectation. I think it is mostly because other languages started this way so JS has to be consistent with them. To me, I think this is not the best thing to do because all the customer gets is an exception message which I think it makes rehydration harder.